source/common/http/conn_manager_impl.cc

```diff
@@ -598,6 +599,17 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(HeaderMapPtr&& headers,
   // called with end_stream=true.
   maybeEndDecode(end_stream);
 
+  if (connection_manager_.config_.validateHeaders()) {
+    ENVOY_STREAM_LOG(trace, "validating header values\n{}", *this, *request_headers_);
+    if (!Http::HeaderUtility::validateHeaders(*request_headers_)) {
```

Comment:
From a perf perspective, this ends up being a full redundant header scan for HTTP/2 as nghttp2 has already validated headers at this point.  It might be worth moving this check in the HTTP/1 codec (maybe even doing onHeaderValue? rather than once at the end) and we could ASSERT that all characters are valid so that as we add more transports (QUIC etc.) we have consistent behavior.

Reply:
+1

Reply:
@alyssawilk If I do it in the http/1.1 codec, it seems to make sense to handle it similarly to [this code](https://github.com/envoyproxy/envoy/blob/5ee70cb19ca3bb6d817bae497b63c2a250d2d16b/source/common/http/http1/codec_impl.cc#L424-L430), which is doing similar albeit more limited, header value checking (it even references the same RFC section!)

Reply:
Yeah, I think with the runtime guard false, it does the old thing, and with the runtime guard true, it does your new thing.
runtime starts out false, we flip it with the release, and then remove old code the following release.
If there's any docs clean up which would make that more clear in future feel free to include it in this PR (with my thanks)